zip-0155.rst

```diff
@@ -0,0 +1,246 @@
+  | ``0x02``   | ``IPV6``    | 16                     | IPv6 address (globally routed internet) |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x04``   | ``TORV3``   | 32                     | Tor v3 hidden service address           |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x05``   | ``I2P``     | 32                     | I2P overlay network address             |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x06``   | ``CJDNS``   | 16                     | Cjdns overlay network address           |
+  +------------+-------------+------------------------+-----------------------------------------+
+
+Network ID ``0x03`` is intentionally not assigned. In BIP 155 [#bip-0155]_ it was
+assigned to Tor v2 hidden service addresses, but those addresses are no longer
+supported by the latest Tor client code, and MUST NOT be used once this ZIP is
+deployed.
+
+Clients SHOULD gossip addresses from all known networks even if they are currently
+not connected to some of them. That could help multi-homed nodes and make it more
+difficult for an observer to tell which networks a node is connected to.
+
+Clients MUST NOT gossip addresses from unknown networks, because they have no means
+to validate those addresses and so can be tricked to gossip invalid addresses.
+
+Further network ID numbers MUST be reserved in a new ZIP.
+
+Clients MUST reject messages that contain addresses that have a different length
+than specified in this table for a specific network ID, as these are meaningless.
+
+Network address encodings
+-------------------------
+
+The ``IPV4`` and ``IPV6`` network IDs use addresses encoded in the usual way for
+binary IPv4 and IPv6 addresses in network byte order (big endian).
+
+Tor v3 address encoding
+'''''''''''''''''''''''
+
+According to the spec [#Tor-rendezvous-v3]_, version 3 ``.onion`` addresses are
+encoded as follows::
+
+  onion_address = base32(PUBKEY | CHECKSUM | VERSION) + ".onion"
+  CHECKSUM = H(".onion checksum" | PUBKEY | VERSION)[:2]
+
+where:
+
+* ``PUBKEY`` is the 32 bytes Ed25519 master pubkey of the hidden service;
+* ``VERSION`` is a one byte version field (default value ``0x03``);
+* ``".onion"`` and ``".onion checksum"`` are constant strings;
+* ``CHECKSUM`` is truncated to two bytes before inserting it in ``onion_address``;
+* ``H()`` is the SHA3-256 cryptographic hash function.
+
+Tor v3 addresses MUST be sent with the ``TORV3`` network ID, with the 32-byte
+``PUBKEY`` part in the address field. As ``VERSION`` will always be '\x03' in the
+case of v3 addresses, this is enough to reconstruct the onion address.
+
+I2P address encoding
+''''''''''''''''''''
+
+Like Tor, I2P naming uses a base32-encoded address format [#I2P-naming]_.
+
+I2P uses 52 characters (256 bits) to represent the full SHA-256 hash, followed by
+``.b32.i2p``.
```

Comment:
Should we mention padding bits?  52 base32 characters is 52*5 = 260 bits, so there's 4 bits of extra information that aren't part of the SHA-256 hash's representation.

Reply:
Padding is part of the base32 spec. I think this is clear enough.

Reply:
```suggestion
``.b32.i2p``. The base32 encoding does not include ``"="`` padding characters.
```